L82-core-event-engine.md

```diff
@@ -0,0 +1,539 @@
+  /// On Endpoint creation, call \a CreateSliceAllocator with the name of the
+  /// endpoint peer (a URI string, most likely).
+  virtual std::unique_ptr<SliceAllocator> CreateSliceAllocator(
+      absl::string_view peer_name) = 0;
+};
+```
+
+Finally, many `EventEngine` pieces rely on `Callback`s, `TaskHandle`s, and
+`ResolvedAddress`es.
+
+```cpp
+class EventEngine {
+ public:
+  /// Basic callable function. The first argument to all callbacks is an
+  /// absl::Status indicating the status of the operation associated with this
+  /// callback. Each EventEngine method that takes a callback parameter, defines
+  /// the expected sets and meanings of statuses for that use case.
+  using Callback = std::function<void(absl::Status)>;
+  /// Callback handle, used to cancel a callback.
+  struct TaskHandle {
+    intptr_t keys[2];
+  };
+  /// Thin wrapper around a platform-specific sockaddr type. A sockaddr struct
+  /// exists on all platforms that gRPC supports.
+  ///
+  /// Platforms are expected to provide definitions for:
+  /// * sockaddr
+  /// * sockaddr_in
+  /// * sockaddr_in6
+  class ResolvedAddress {
+   public:
+    static constexpr socklen_t MAX_SIZE_BYTES = 128;
+
+    ResolvedAddress(const sockaddr* address, socklen_t size);
+    ResolvedAddress() = default;
+    ResolvedAddress(const ResolvedAddress&) = default;
+    const struct sockaddr* address() const;
+    socklen_t size() const;
+
+   private:
+    char address_[MAX_SIZE_BYTES];
+    socklen_t size_ = 0;
+  };
+  virtual ~EventEngine() = default;
+```
+
+### Endpoints
+
+```cpp
+  /// One end of a connection between a gRPC client and server. Endpoints are
+  /// created when connections are established, and Endpoint operations are
+  /// gRPC's primary means of communication.
+  ///
+  /// Endpoints must use the provided SliceAllocator for all data buffer memory
+  /// allocations. gRPC allows applications to set memory constraints per
+  /// Channel or Server, and the implementation depends on all dynamic memory
+  /// allocation being handled by the quota system.
+  class Endpoint {
+   public:
+    /// Shuts down all connections and invokes all pending read or write
+    /// callbacks with an error status.
+    virtual ~Endpoint() = default;
+    /// Reads data from the Endpoint.
+    ///
+    /// When data is available on the connection, that data is moved into the
+    /// \a buffer, and the \a on_read callback is called. The caller must ensure
+    /// that the callback has access to the buffer when executed later.
+    /// Ownership of the buffer is not transferred. Valid slices *may* be placed
+    /// into the buffer even if the callback is invoked with a non-OK Status.
+    ///
+    /// For failed read operations, implementations should pass the appropriate
+    /// statuses to \a on_read. For example, callbacks might expect to receive
+    /// CANCELLED on endpoint shutdown.
+    virtual void Read(Callback on_read, SliceBuffer* buffer) = 0;
```

Comment:
We need to make the contract clear here that there can be _at most one_ outstanding read on each `Endpoint` at any given time. Calls to `Read` while a read is already in flight (i.e. the `on_read` callback has not been executed yet for some previous call to `Read`) should result in an error. We could change the API to `virtual absl::Status Read(...)` and return a friendly error synchronously. We could immediately call their callback with some error status we define here. We could state that the EventEngine implementation should abort, given that something is using the API improperly and needs to be fixed. @ctiller @markdroth @nicolasnoble thoughts? I'm leaning towards the 2nd option, executing the callback immediately with an error status; I think I'd prefer the first option if we had other uses of returning an immediate error.

Reply:
Discussing with @markdroth, @ctiller, and @nicolasnoble, it seems best to abort in this situation. Clients that are using the API incorrectly are unlikely to recover gracefully. I've added that language to the documentation.